File: questModule/src/main/java/ru/nekit/android/qls/quest/answer/MetricsQuestAnswerChecker.java
package ru.nekit.android.qls.quest.answer;

import android.support.annotation.NonNull;

import ru.nekit.android.qls.quest.IQuest;
import ru.nekit.android.qls.quest.answer.shared.QuestAnswerChecker;
import ru.nekit.android.qls.quest.math.MathematicalSignComparison;
import ru.nekit.android.qls.quest.types.MetricsQuest;
import ru.nekit.android.qls.quest.types.NumberSummandQuest;

public class MetricsQuestAnswerChecker extends QuestAnswerChecker<MathematicalSignComparison> {

    @Override
    public boolean checkAlternativeInput(@NonNull IQuest quest,
                                         @NonNull MathematicalSignComparison answer) {
        NumberSummandQuest numberSummandQuest = (NumberSummandQuest) quest;
        switch (quest.getQuestionType()) {

            case COMPARISON:

                int leftMetricsSum = getMetricsSum(numberSummandQuest.leftNode);
                int rightMetricsSum = getMetricsSum(numberSummandQuest.rightNode);
                MathematicalSignComparison metricsSign = MathematicalSignComparison.EQUAL;
                if (leftMetricsSum > rightMetricsSum) {
                    metricsSign = MathematicalSignComparison.GREATER;
                } else if (leftMetricsSum < rightMetricsSum) {
                    metricsSign = MathematicalSignComparison.LESS;
                }
                return metricsSign == answer;

            default:

                return false;

        }
    }

    private int getMetricsSum(int[] node) {
        int metricsSum = 0;
        int[] converterValues = new int[]{MetricsQuest.METRICS_CENTIMETER_IN_METER,
                MetricsQuest.METRICS_CENTIMETER_IN_DECIMETER, 1};
        for (int i = 0; i < MetricsQuest.METRICS_ITEM_COUNT; i++) {
            metricsSum += converterValues[i] * node[i];
        }
        return metricsSum;
    }

    @Override
    public boolean checkStringInput(@NonNull IQuest quest, @NonNull String value) {
        NumberSummandQuest numberSummandQuest = (NumberSummandQuest) quest;
        switch (quest.getQuestionType()) {
            case SOLUTION:
                return getMetricsSum(numberSummandQuest.leftNode) == Integer.valueOf(value);
            default:
                return false;
        }
    }
}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/answerChecker/MetricsQuestAnswerChecker.kt
package ru.nekit.android.qls.domain.answerChecker

import ru.nekit.android.qls.domain.answerChecker.common.QuestAnswerChecker
import ru.nekit.android.qls.domain.model.math.MathematicalSignComparison
import ru.nekit.android.qls.domain.model.quest.MetricsQuest
import ru.nekit.android.qls.domain.model.quest.NumberSummandQuest
import ru.nekit.android.qls.shared.model.QuestionType

class MetricsQuestAnswerChecker : QuestAnswerChecker<NumberSummandQuest>() {

    override fun checkAlternativeInput(quest: NumberSummandQuest,
                                       answer: Any): Boolean {
        return answer == when (quest.questionType) {

            QuestionType.COMPARISON -> {

                val leftMetricsSum = quest.leftNodeSum
                val rightMetricsSum = quest.rightNodeSum
                when {
                    leftMetricsSum > rightMetricsSum -> MathematicalSignComparison.GREATER
                    leftMetricsSum < rightMetricsSum -> MathematicalSignComparison.LESS
                    else -> MathematicalSignComparison.EQUAL
                }
            }

            else -> null

        }
    }

    private fun getMetricsSum(node: IntArray): Int {
        val converterValues = intArrayOf(MetricsQuest.METRICS_CENTIMETER_IN_METER,
                MetricsQuest.METRICS_CENTIMETER_IN_DECIMETER,
                1)
        return (0 until MetricsQuest.METRICS_ITEM_COUNT).sumBy { converterValues[it] * node[it] }
    }

    override fun checkStringInput(quest: NumberSummandQuest, value: String): Boolean {
        return when (quest.questionType) {
            QuestionType.SOLUTION -> getMetricsSum(quest.leftNode) == Integer.valueOf(value)
            else -> false
        }
    }
}